executor/src/witgen/multiplicity_column_generator.rs

```diff
@@ -0,0 +1,202 @@
+use std::collections::{BTreeMap, HashMap};
+
+use powdr_ast::{
+    analyzed::{AlgebraicExpression, Identity, PolynomialType, SelectedExpressions},
+    parsed::visitor::AllChildren,
+};
+use powdr_executor_utils::expression_evaluator::{ExpressionEvaluator, OwnedTerminalValues};
+use powdr_number::FieldElement;
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+
+use crate::witgen::machines::{
+    profiling::{record_end, record_start},
+    Connection,
+};
+
+use super::FixedData;
+
+static MULTIPLICITY_WITGEN_NAME: &str = "multiplicity witgen";
+
+pub struct MultiplicityColumnGenerator<'a, T: FieldElement> {
+    fixed: &'a FixedData<'a, T>,
+}
+
+impl<'a, T: FieldElement> MultiplicityColumnGenerator<'a, T> {
+    pub fn new(fixed: &'a FixedData<'a, T>) -> Self {
+        Self { fixed }
+    }
+
+    /// Takes a map of witness columns and extends it with the multiplicity columns
+    /// reference in the phantom lookups.
+    pub fn generate(
+        &self,
+        columns: HashMap<String, Vec<T>>,
+        publics: BTreeMap<String, Option<T>>,
+    ) -> HashMap<String, Vec<T>> {
+        record_start(MULTIPLICITY_WITGEN_NAME);
+
+        log::trace!("Starting multiplicity witness generation.");
+        let start = std::time::Instant::now();
+
+        // Several range constraints might point to the same target
+        let mut multiplicity_columns = BTreeMap::new();
+
+        let phantom_lookups = self
+            .fixed
+            .analyzed
+            .identities
+            .iter()
+            .filter_map(|identity| match identity {
+                // TODO(bus_interaction)
+                Identity::PhantomLookup(_) => Some(Connection::try_from(identity).unwrap()),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+
+        let all_columns = columns
+            .into_iter()
+            .map(|(name, col)| {
+                let poly_id = self.fixed.try_column_by_name(&name).unwrap();
+                (poly_id, col)
+            })
+            .chain(
+                self.fixed
+                    .fixed_cols
+                    .iter()
+                    // TODO: Avoid clone
```

Comment:
I think this should not be too hard `OwnedTerminalValues` could do everything it does with a reference, but that's a different PR.